src/oic/oic/provider.py

```diff
@@ -2049,3 +1883,449 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
+        Remove Session IDs from the session DB.
+
+        :param usids: List of session IDs
+        """
+        _sdb = self.sdb
+        # Clean out all sessions
+        for sid in usids:
+            del _sdb[sid]
+
+    def logout_all_clients(self, sid: str) -> Dict:
+        """
+        Collect information necessary to logout one user from all clients he/she has been using.
+
+        :param sid: Session ID
+        :return: Dictionary with logout information
+        """
+        _sdb = self.sdb
+
+        uid = _sdb.get_uid_by_sid(sid)
+        _client_sid = {}
+        # Find all the session IDs this user has gotten
+        usids = session_get(_sdb, "uid", uid)
+        # Find all RPs this user has logged it from
+        for usid in usids:
+            _client_sid[_sdb[usid]["client_id"]] = usid
+
+        # Front-/Backchannel logout ?
+        _cdb = self.cdb
+        _iss = self.name
+        bc_logouts = {}
+        fc_iframes = {}
+        for _cid, _csid in _client_sid.items():
+            if "backchannel_logout_uri" in _cdb[_cid]:
+                _sub = _sdb[_csid]["sub"]
+                _spec = self.do_back_channel_logout(_cdb[_cid], _sub, _csid)
+                if _spec:
+                    bc_logouts[_cid] = _spec
+            elif "frontchannel_logout_uri" in _cdb[_cid]:
+                # Construct an IFrame
+                _iframe = do_front_channel_logout_iframe(_cdb[_cid], _iss, _csid)
+                if _iframe:
+                    fc_iframes[_cid] = _iframe
+
+        # Clean out all sessions
+        self.clean_sessions(usids)
+
+        return {"blu": bc_logouts, "flu": fc_iframes}
+
+    def logout_from_client(self, sid: str, client_id: str) -> Dict:
+        """
+        Collect information necessary to log out from client.
+
+        Note that if a client has both back channel and front channel logout registered both
+        will be attempted.
+
+        :param sid: Session ID
+        :param client_id: Client ID
+        :return: Dictionary with back_channel and front_channel logout info.
+        """
+        _cdb = self.cdb
+        _sso_db = self.sdb
+
+        res = {
+            "blu": {},
+            "flu": {},
+        }  # type: Dict[str, Dict[str, Union[None, str, Tuple[str,str]]]]
+
+        if "backchannel_logout_uri" in _cdb[client_id]:
+            _sub = _sso_db[sid]["sub"]
+            _spec = self.do_back_channel_logout(_cdb[client_id], _sub, sid)
+            if _spec:
+                res["blu"] = {client_id: _spec}
+        elif "frontchannel_logout_uri" in _cdb[client_id]:
+            # Construct an IFrame
+            _iframe = do_front_channel_logout_iframe(_cdb[client_id], self.name, sid)
+            if _iframe:
+                res["flu"] = {client_id: _iframe}
+
+        # Clean out sessions
+        self.clean_sessions([sid])
+
+        return res
+
+    def end_session_endpoint(
+        self, request: str = "", cookie: Union[str, SimpleCookie, None] = None, **kwargs
```

Comment:
Again, isn't this `Optional[Union[str, SimpleCookie]]`?

Reply:
OK